ambry-account/src/main/java/com/github/ambry/account/CompositeAccountService.java

```diff
@@ -0,0 +1,191 @@
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.utils.Utils.*;
+
+
+/**
+ * An implementation of {@link AccountService} that combines two sources. The "primary" acts as the source of
+ * truth for users of this class, but results obtained are compared against the "secondary" source. This class can be
+ * useful for safe migrations between account service implementations using different backing stores.
+ */
+public class CompositeAccountService implements AccountService {
+
+  private static final Logger logger = LoggerFactory.getLogger(CompositeAccountService.class);
+  private static final String ACCOUNT_DATA_CONSISTENCY_CHECKER_PREFIX = "account-data-consistency-checker";
+  private final AccountService primaryAccountService;
+  private final AccountService secondaryAccountService;
+  private final AccountServiceMetrics accountServiceMetrics;
+  private final CompositeAccountServiceConfig config;
+  private final ScheduledExecutorService scheduler;
+  private static long getAccountRequestCount = 0;
+  private static long getContainerRequestCount = 0;
+
+  public CompositeAccountService(AccountService primaryAccountService, AccountService secondaryAccountService,
+      AccountServiceMetrics accountServiceMetrics, CompositeAccountServiceConfig config) {
+    this.primaryAccountService = primaryAccountService;
+    this.secondaryAccountService = secondaryAccountService;
+    this.accountServiceMetrics = accountServiceMetrics;
+    this.config = config;
+    scheduler = Utils.newScheduler(1, ACCOUNT_DATA_CONSISTENCY_CHECKER_PREFIX, false);
+    scheduler.scheduleAtFixedRate(this::compareAccountMetadata, config.consistencyCheckerIntervalMinutes,
+        config.consistencyCheckerIntervalMinutes, TimeUnit.MINUTES);
+  }
+
+  @Override
+  public Account getAccountById(short accountId) {
+    Account primaryResult = primaryAccountService.getAccountById(accountId);
+    Account secondaryResult = secondaryAccountService.getAccountById(accountId);
+    if ((++getAccountRequestCount % 100 < config.samplingPercentageForGetAccountConsistencyCheck)
+        && primaryResult != null && !primaryResult.equals(secondaryResult)) {
+      logger.warn("Inconsistency detected between primary and secondary for accountId ={}", accountId);
+      accountServiceMetrics.getAccountDataInconsistencyCount.inc();
+    }
+    return primaryResult;
+  }
+
+  @Override
+  public Account getAccountByName(String accountName) {
+    Account primaryResult = primaryAccountService.getAccountByName(accountName);
+    Account secondaryResult = secondaryAccountService.getAccountByName(accountName);
+    if (++getAccountRequestCount % 100 < config.samplingPercentageForGetAccountConsistencyCheck && primaryResult != null
+        && !primaryResult.equals(secondaryResult)) {
+      logger.warn("Inconsistency detected between primary and secondary for accountName ={}", accountName);
+      accountServiceMetrics.getAccountDataInconsistencyCount.inc();
+    }
```

Comment:
It seems `compareAccountMetadata` method does an exhaustive check periodically. Any specific reason to do a sampling check?

Reply:
That was my idea, and now I'm having second thoughts about it too.